src/primitives/transaction.cpp

```diff
@@ -237,142 +266,318 @@ std::string CTxCertifierLockOut::ToString() const
+        }
+
+        if (joinsplit.vpub_old > MAX_MONEY) {
+            return state.DoS(100, error("CheckOutputAmounts(): joinsplit.vpub_old too high"),
+                             REJECT_INVALID, "bad-txns-vpub_old-toolarge");
+        }
+
+        if (joinsplit.vpub_new > MAX_MONEY) {
+            return state.DoS(100, error("CheckOutputAmounts(): joinsplit.vpub_new too high"),
+                             REJECT_INVALID, "bad-txns-vpub_new-toolarge");
+        }
+
+        if (joinsplit.vpub_new != 0 && joinsplit.vpub_old != 0) {
+            return state.DoS(100, error("CheckOutputAmounts(): joinsplit.vpub_new and joinsplit.vpub_old both nonzero"),
+                             REJECT_INVALID, "bad-txns-vpubs-both-nonzero");
+        }
+
+        nCumulatedValueOut += joinsplit.vpub_old;
+        if (!MoneyRange(nCumulatedValueOut)) {
+            return state.DoS(100, error("CheckOutputAmounts(): txout total out of range"),
+                             REJECT_INVALID, "bad-txns-txouttotal-toolarge");
+        }
+    }
+
+    return true;
 }
 
-CAmount CTransaction::GetJoinSplitValueIn() const
+bool CTransactionBase::CheckInputsDuplication(CValidationState &state) const
 {
-    CAmount nValue = 0;
-    for (std::vector<JSDescription>::const_iterator it(vjoinsplit.begin()); it != vjoinsplit.end(); ++it)
+    // Check for duplicate inputs
+    std::set<COutPoint> vInOutPoints;
+    for(const CTxIn& txin: GetVin())
     {
-        // NB: vpub_new "gives" money to the value pool just as inputs do
-        nValue += it->vpub_new;
+        if (vInOutPoints.count(txin.prevout))
+            return state.DoS(100, error("CheckInputsDuplications(): duplicate inputs"),
+                             REJECT_INVALID, "bad-txns-inputs-duplicate");
+        vInOutPoints.insert(txin.prevout);
+    }
 
-        if (!MoneyRange(it->vpub_new) || !MoneyRange(nValue))
-            throw std::runtime_error("CTransaction::GetJoinSplitValueIn(): value out of range");
+    // Check for duplicate joinsplit nullifiers in this transaction
+    std::set<uint256> vJoinSplitNullifiers;
+    for(const JSDescription& joinsplit: GetVjoinsplit())
+    {
+        for(const uint256& nf: joinsplit.nullifiers)
+        {
+            if (vJoinSplitNullifiers.count(nf))
+                return state.DoS(100, error("CheckInputsDuplications(): duplicate nullifiers"),
+                             REJECT_INVALID, "bad-joinsplits-nullifiers-duplicate");
+
+            vJoinSplitNullifiers.insert(nf);
+        }
     }
 
-    return nValue;
+    return true;
 }
 
-double CTransaction::ComputePriority(double dPriorityInputs, unsigned int nTxSize) const
+bool CTransactionBase::CheckInputsInteraction(CValidationState &state) const
 {
-    nTxSize = CalculateModifiedSize(nTxSize);
-    if (nTxSize == 0) return 0.0;
+    if (IsCoinBase())
+    {
+        // There should be no joinsplits in a coinbase transaction
+        if (GetVjoinsplit().size() > 0)
```

Comment:
We MUST exclude also all the cross chain outputs. Moreover the checks for coinbases txs are applicable only for CTransactions. Make this method virtual and implement it differently for certificates and transactions

Reply:
Done